### PR TITLE
feat: basic implement for RESP parser

### PR DESCRIPTION
**Pull request by DCjanus**
# Feature

+ Zero copy
+ Non-backtracking
+ Type safe

### CODE REVIEW

File: src/buffer.rs [hunk 1]
```diff
@@ -0,0 +1,135 @@
+use std::{cmp::min, io};
```
Comment (wayslog) [inline src/buffer.rs:1]: plz use `bytes::BytesMut`

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,337 @@
+// TODO integer overflow check
```
Comment (wayslog) [inline src/lib.rs:1]: TODO...

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+
+#[derive(Debug, Default)]
+struct IntegerParser(i64);
+
+impl IntegerParser {
+    fn parse(mut self, buffer: &mut Buffer) -> ParseOutput {
+        for i in buffer.iter() {
+            match i {
+                x @ b'0'..=b'9' => self.0 = self.0 * 10 + i64::from(x - b'0'),
+                b'-' => self.0 = -self.0,
```
Comment (wayslog) [inline src/lib.rs:196]: you call parse with default `IntegerParser`, but the initial value is `0`. And the negative value of zero is also zero!
so, if the data is `:-20\r\n`, you  will get a parsed integer as `20`, but not `-20`.

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+#[derive(Debug, Default)]
+struct BulkSizeParser(isize);
+
+impl BulkSizeParser {
+    fn parse(mut self, buffer: &mut Buffer) -> ParseOutput {
+        for i in buffer.iter() {
+            match i {
+                x @ b'0'..=b'9' => self.0 = self.0 * 10 + isize::from(x - b'0'),
+                b'-' => self.0 = -self.0,
```
Comment (wayslog) [inline src/lib.rs:217]: bulk string's size gets only one way to negative: the whole bulk is `$-1\r\n`. So you should make it faster or test it.

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+}
+
+#[derive(Debug, Default)]
+struct ArraySizeParser(usize);
+
+impl ArraySizeParser {
+    fn parse(mut self, buffer: &mut Buffer) -> ParseOutput {
+        for i in buffer.iter() {
+            match i {
+                x @ b'0'..=b'9' => self.0 = self.0 * 10 + usize::from(x - b'0'),
```
Comment (wayslog) [inline src/lib.rs:268]: you  should check if the array is `NULL array`, with content is `*-1\r\n`

File: src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+    fn parse(self, buffer: &mut Buffer) -> ParseOutput {
+        if buffer.iter().position(|x| x == b'\n') == None {
+            return self.into_enum();
+        }
+
+        let mut data = buffer.pop_scanned_buffer();
+        if data.pop() != Some(b'\n') {
+            return Error::ContentWrong.into_enum();
+        }
+        if data.pop() != Some(b'\r') {
```
Comment (wayslog) [inline src/lib.rs:332]: inline command always ends with `\n` but not usually ends with `\r\n`. That's not error.

File: tests/basic.rs [hunk 7]
```diff
@@ -0,0 +1,45 @@
+use resp_parser::{Buffer, Message, RespParser, Result};
+use std::io::Write;
+
```
Comment (wayslog) [inline tests/basic.rs:3]: need more test case...

**Review by wayslog — CHANGES_REQUESTED** [review]
need more benchmarks and test cases

**Comment by wayslog** [comment]
另外我还发现一个问题：如果是 Array 套 Array 的情况你这里没考虑啊。。。

**Comment by DCjanus** [comment]
My code is shit, close this PR

**Pull request closed without merging**